packages/provider-utils/src/resolve.ts

```diff
@@ -0,0 +1,24 @@
+export type Resolvable<T> =
+  | T // Raw value
+  | Promise<T> // Promise of value
+  | (() => T) // Function returning value
+  | (() => Promise<T>); // Function returning promise of value
+
+/**
+ * Resolves a value that could be a raw value, a Promise, a function returning a value,
+ * or a function returning a Promise.
+ */
+export async function resolve<T>(value: Resolvable<T>): Promise<T> {
+  // If it's a function, call it to get the value/promise
+  if (typeof value === 'function') {
+    const result = (value as Function)();
+    return Promise.resolve(result);
+  }
+
+  // Otherwise just resolve whatever we got (value or promise)
+  return Promise.resolve(value);
+}
+
+// Helper for common case of resolving headers
+export type ResolvableHeaders = Resolvable<Record<string, string | undefined>>;
+export const resolveHeaders = (headers: ResolvableHeaders) => resolve(headers);
```

Comment:
I don't know if `ResolvableHeaders` is needed. Do we have a `Headers` type somewhere (warning: might conflict with JS built-ins)

`resolveHeaders` is unnecessary, type inference will automatically do the right thing with just resolve.

Reply:
(if `ResolvableHeaders` is needed, consider moving it to a separate file)

Reply:
Thought about it more: `ResolvableHeaders` has little value over `Resolvable<Headers>`. We should not include it imo (and instead think about a `Headers` type).